### ISSUE TITLE
Question on White Balance Handling in 16C Simulation Pipeline

### ISSUE THREAD

**Issue by excllent123** [SamsungLabs/unified-demosaicing#6]
Dear Dataset Author,

I am writing to seek your advice on a white balance (WB) decision in our 16C simulation pipeline, which we have built on top of your dataset.

**Background**
Our demosaic network was originally trained on DIV2K-based simulated data, following the pipeline: sRGB → inverse gamma → 16C / 16C mosaic. In this pipeline, AWB is implicitly applied by the ISP before the sRGB stage, so the linear domain input to the network is always chromatically neutral.

We have since moved to a new dataset consisting of real Linear RAW captures (uint16, 14-bit) acquired exclusively under D-illuminant, without any white balance applied. Our current 16C simulation operates directly on the raw linear domain. 

Observed Issue
Performance on D-illuminant test scenes is acceptable. However, on A-illuminant test scenes, we observe strong periodic color grid artifacts along high-contrast edges and in overexposed regions.  （DIV2K vs  Hard Dataset ）

Our Questions
1. In your intended use of the dataset, is WB normalization expected to be applied before 16C simulation, or is the network expected to operate on raw unbalanced channels?

2. If WB should be applied internally during simulation (i.e., normalize channels for the warp step, then save back to the RAW domain), do you recommend using a fixed D-illuminant gain (e.g., the camera's stored CCT gains), or should per-image illuminant estimation be performed?

3. Does your dataset include captures under multiple illuminants (e.g., A, TL, D), or only D-illuminant? If multi-illuminant data exists, we would be very interested in accessing it to improve generalization.

4. Would you recommend conditioning the network on WB gains (r_gain, b_gain) as auxiliary metadata, as an alternative or complement to normalizing in the simulation stage?

Our current working hypothesis is that WB normalization should be applied internally during simulation — neutralizing the channels before the warp step — while noise residuals are still extracted from the original unbalanced RAW captures, so that the final saved data remains in the true sensor RAW domain. We would greatly appreciate your perspective on whether this approach aligns with the intended design of the dataset.

Thank you very much for your time and for making this dataset available. We look forward to your reply.

Best regards

**Comment by tedlasai** [comment]
Hi hope you are well.

Let me try to answer generally and then specifically.  First I am usure what 16C stands for here (16 channels?) This is a 3 channel dataset.

Our dataset is a raw space dataset captured under a single illuminant. The demosaicing problem is fundamentally about taking a raw mosaic and interpolating the gaps (the other 2 color channels per pixel). It is the first step you would do when you capture a mosaic. Then any processing after such as wb, warping, etc can easily be done on the 3 channel image. In principle, there is no reason for the demosaicing network to know about the white point. It is primarily an interpolation procedure.

What could be happening is that training the demosaicing network under a specific illuminant could mean that it doesn't see certain colors/patterns as often in its training data that appear under other illuminants.  A simple augmentation, could be done by applying different white balances to the raw training data. This would approximately simulate different raws under other white points. 


Your post makes it unclear what you are doing with WB normalization.... But one thing that is clear is that your previous approach of using DIV2K and undoing gamma is incorrect. Note that the sRGB gamma is just one of many non-linearities (it is a perceptual encoding non-linearity introduced to prioritize darker regions). However, from what I know DIV2K could contain images with other tone-mappings and non-linearities in the pipeline. Most likely the data is _not linear_ just by inverting the gamma. 

In your visualization it is unclear what you mean by DIV2K vs Hard dataset? What are the training/testing splits? Are you training on HDD and testing on Div2K? If so, there are many issues with this. Training on RAW and testing on non-linear data in DIV2K is bound to cause issues.

**Comment by tedlasai** [comment]
Sorry, I didn't answer all your specific questions as I didn't understand context. But feel free to elaborate and hopefully we can help you find a solution!

**Comment by excllent123** [comment]
First, I apologize for the unclear description. Let me provide more details:

- **16C** refers to the Nona-Bayer Pattern, specifically the 4×4 (Nona) model. We sample the data into a 4×4 pattern as input to the network, and the output is the corresponding Bayer RAW data.
- **Training set**: HDD Dataset. We sample it into a 4×4 pattern input, and the ground truth (GT) is the corresponding Bayer RAW.
- **Test set**: Real RAW data captured by a sensor, also with a 4×4 Nona-Bayer pattern.

To compare HDD and DIV2K, we train the same network on each. The pipelines are as follows:

- **DIV2K-based simulated data**: sRGB → inverse gamma → 16C mosaic. The GT is the extracted Bayer RAW. Since DIV2K images are already white-balanced, we apply WB to the input and inverse WB to the output during testing.
- **HDD data pipeline**: directly sample to 4×4 pattern input, GT is the corresponding Bayer RAW. No preprocessing is applied to the input during testing.

The results (shown in the attached figure) indicate that under illuminant D, the performance with HDD is significantly better than with DIV2K. However, under illuminant A (as seen in the figure), the performance degrades severely.

**A Light Result** (DIV2K vs HDD)
<img width="2560" height="1380" alt="Image" src="https://github.com/user-attachments/assets/a8582644-c510-407c-b5b4-8a2ce6336fb9" />

**D Light Result** (DIV2K vs HDD)
<img width="2560" height="1380" alt="Image" src="https://github.com/user-attachments/assets/41d0d312-aec0-40da-a1dc-dae7b87c3e9f" />

In summary, the issue is that the HDD dataset contains only a single illuminant. To improve this situation, do you have any suggestions? Or how can I augment data under other illuminant conditions? Thank you!

**Comment by tedlasai** [comment]
Ok this helps, thanks!

First, I think you are confusing Nona meaning. We did Nona, which is a 3x3 repeating pattern (screenshot below). I am trying to clarify the definitions so that way I know you are working on the same thing :) Nona is always 3x3 (as nona means 9) . 

Second, you are training and testing on 2 different sensors. Demosaicing is a sensor-specific process... More crucially than lighting is that the demosaicing is typically a sensor-specific model (like in our case). Meaning  that your case of training on our data and testing on "real data" from another sensor is invalid. 

Third, DIV2K-based simulated data: sRGB → inverse gamma → 16C mosaic. The GT is the extracted Bayer RAW. Does DIV2K provide bayer raw? if so, why don't you just sample mosaics from that?

Fourth, yes there are many ways to augment raw datasets including ours. A simple thing is to sample white points, divide out the base illuminant by applying WB (diagonal correction). Then adding a color cast by applying another illuminant (doing the opposite of WB, i.e, inverse diagonal correction). We do something like this here (although this is for sRGB to Raw mapping, but the same idea could be used for raw mapping) -> https://openaccess.thecvf.com/content/ICCV2023/papers/Seo_Graphics2RAW_Mapping_Computer_Graphics_Images_to_Sensor_RAW_Images_ICCV_2023_paper.pdf 

Here sample white points from a bunch of DNGs, then we can choose an arbitrary white point to render back to raw. You can do this by taking Raw image (under D from our dataset) -> WB and to so its raw Image (without color cast) -> then apply inverse WB to simulate another illuminant. 

A more thorough option would be to do (this also handles color balancing)

Raw image (under D from our dataset) -> WB + CST and to so its XYZ Image-> then apply inverse CST and inverse WB to simulate another illuminant.

Generally, I would refer you to read both our paper and UPI (both are srgb to raw mapping papers). A lot of the ideas apply for raw to raw mappings as well. 

UPI - https://arxiv.org/abs/1811.11127


Hope this makes sense.



<img width="571" height="379" alt="Image" src="https://github.com/user-attachments/assets/9fda3986-bfe4-4748-8483-d1d1786360b1" />

**Comment by excllent123** [comment]
Sorry, I misunderstood the meaning of Nona. It should actually be **hexadeca Bayer**, with the pattern shown in the attached image. 

<img width="689" height="344" alt="Image" src="https://github.com/user-attachments/assets/ec6bd9fe-12da-4347-bd44-ca97aa2bbe84" />

More details about the pattern can be found here:  https://arxiv.org/pdf/2604.23268

Regarding the second point, as stated in the paper(https://arxiv.org/pdf/2604.23268 ), it is indeed feasible to train on simulated data and test on real sensor data.

Third, DIV2K does not provide Bayer RAW images, so we need to sample to obtain the mosaic images.

As for the fourth point, we will try the method you suggested. Thank you.

Best regards

**Comment by tedlasai** [comment]
2. Sure you can do it, but technically there is a domain gap here that is not being bridged here. 

Sounds good thanks, I am closing this issue, but feel free to open this back up if you need help.
